### ISSUE TITLE
The development tree produces hanging, 100%cpu salt-master processes

### ISSUE THREAD

**Issue by GreatSnoopy** [saltstack/salt#26366]
I followed the instructions at 
http://docs.saltstack.com/en/latest/topics/development/hacking.html
to get a separate salt environment for tests, debugging, extending and so on.
Used on ubuntu 14.04, set up virtualenv.
However, the current code seems to produce cpu eating salt-masters as soon as the first execution module is invoked. It should be noted that until some minion interaction is required, everything idles fine. 
stracing the process eating cpu leads to 
```C
poll([{fd=13, events=POLLIN}], 1, 0)    = 0 (Timeout)
poll([{fd=15, events=POLLIN}], 1, 0)    = 0 (Timeout)

poll([{fd=13, events=POLLIN}, {fd=15, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=13, events=POLLIN}], 1, 0)    = 0 (Timeout)
poll([{fd=15, events=POLLIN}], 1, 0)    = 0 (Timeout)

poll([{fd=13, events=POLLIN}, {fd=15, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=13, events=POLLIN}], 1, 0)    = 0 (Timeout)
poll([{fd=15, events=POLLIN}], 1, 0)    = 0 (Timeout)



poll([{fd=13, events=POLLIN}, {fd=15, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=13, events=POLLIN}], 1, 0)    = 0 (Timeout)
poll([{fd=15, events=POLLIN}], 1, 0)    = 0 (Timeout)
```
In very fast succession. Forcing a stack trace to the python process leads eventually to 
salt/utils/event.py
And more exactly at this :
```python
        while not wait or time.time() <= timeout_at:
            try:
                # convert to milliseconds
                socks = dict(self.poller.poll(wait * 1000))
                if socks.get(self.sub) != zmq.POLLIN:
                    continue
```
It seems that the code enters on the `socks.get(self.sub) != zmq.POLLIN` branch which then goes on and on forever (I put a debug statement before continue, and it fires forever).
I am not yet sure why this never ends and why even for a few seconds this fires so rapidly (I dont suppose its normal to hog the cpu even for a a few seconds with which wait eventually is called).
Maybe someone working on the transport/zmq core could take over from here and recheck  the logic of the above code, something seems not right there.
In the current stable code, the construct has that POLLIN check outside the try block, and this is why maybe it does not loop forever due to continue :
```python
        while not wait or time.time() <= timeout_at:
            # convert to milliseconds
            socks = dict(self.poller.poll(wait * 1000))
            if socks.get(self.sub) != zmq.POLLIN:
                continue

            try:
                # Please do not use non-blocking mode here.
                # Reliability is more important than pure speed on the event bus.
                ret = self.get_event_block()
```

**Comment by jfindlay** [comment]
@GreatSnoopy, thanks for the report.

**Comment by pitatus** [comment]
I've been seeing this since about Thursday of last week:

**Comment by cachedout** [comment]
My first thought is that this may be related to the event caching code that
was merged late last week.

**Comment by GreatSnoopy** [comment]
@jfindlay you're welcome. I'll submit another one soon :) (  -d does not work, complains at fork time)

**Comment by waynew** [comment]
I think this is the same issue as #26301, I had the same issue on Arch. I'll be watching for a potential fix for this :+1:

**Comment by GreatSnoopy** [comment]
I was also looking into it out of curiosity, however I'm still not able to figure out what's wrong, the issue could be a little bit deeper and just manifesting itself here.
What I was able to get is that  the actual cpu hogging really happens here :


        while not wait or time.time() <= timeout_at:
            # convert to milliseconds
            socks = dict(self.poller.poll(wait * 1000))
            if socks.get(self.sub) != zmq.POLLIN:
                continue

This piece of code gets executed like crazy and this actually eats the CPU. 
This eventually expires (after 5 seconds, i think) but after this a new cycle starts again, as _get_event is called again.
However, the code  seems seems quite similar  as in the current Centos/RHEL EPEL build. I do not know if  It's supposed to run just a few times  and eat cpu in that loop for a few 5 seconds rounds or if that even inside that loop it should not iterate so fast.

**Comment by waynew** [comment]
I don't know *why* what's happening is happening, but based on my knowledge of polling what's going on here is that something is saying, "Oh yeah, there's data here" and so the polling reads... nothing?

And because there's nothing to read, there's nothing to process, so it takes nearly zero time and spins back through hitting the poll call (again) and... well, does the same thing. Very quickly. Over and over again.

Obviously this is The Wrong Thing™. I find it interesting that it doesn't seem to affect other platforms? At least AFAIK it's only been reported under Arch and Ubuntu.

**Comment by GreatSnoopy** [comment]
I expect to find it on other distributions, as well - should I try the current development tree.

**Comment by cachedout** [comment]
I can only confirm this on the develop branch and not on the 2015.8 branch. Is that the same for others who have seen this issue?

**Comment by waynew** [comment]
I thought that I may have used https://github.com/saltstack/salt/releases/tag/v2015.8.0rc3 and done a `pip install <tarfile>` to test at least once. I'm not seeing that in my history, though. But I am definitely seeing `sudo pacman -U salt-git-v2015.8.0rc3.1192.g909556a-1-any.pkg.tar.xz`, so at the very least I definitely installed it from the [AUR](https://aur.archlinux.org/packages/salt-git/) PKGBUILD.

**Comment by cachedout** [comment]
Please try the fix in #26443

**Comment by cachedout** [comment]
Per the comments in #26443, I'm fairly sure this is now resolved.
